fix: Gitlab scm_provider - don't create transport from scratch (#15424)

an-toine opened the pull request:
Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [X] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Since ArgoCD 2.8.0, the Gitlab scm_provider seems to ignore proxy settings of pod argocd-applicationset-controller.

Instead of creating an `http.Transport` from scratch, this PR aims at cloning the default one to inherit from proxy settings read from environment variables.

Closes #15424

Antoine

an-toine commented:
> This is a good idea, miss other usages such as
> 
>     * https://github.com/argoproj/argo-cd/blob/master/applicationset/services/pull_request/gitlab.go#L35
> 
> 
> what about github ?

I've implemented the same change for the Gitlab `pull_request` provider. I've also updated both providers for Gitea.

Regarding Github, I can't test if the bug is present, and telling it only from the code is a bit beyond my Go skills 🙂 

Antoine

ishitasequeira reviewed (changes requested):
Thanks for the fix @an-toine. The changes LGTM.
can you fix the lint failure and rebase your branch?

an-toine commented:
@ishitasequeira :
  * I have rebased the PR against master
  * The lint failure seems to be caused by a caching issue. I don't have enough permissions to clear it though, can you take care of this ?

Antoine

ishitasequeira commented:
> @ishitasequeira :
> 
> * I have rebased the PR against master
> * The lint failure seems to be caused by a caching issue. I don't have enough permissions to clear it though, can you take care of this ?
> 
> Antoine

I don't think I have write access to your branch. Can you provide me access to it?

an-toine commented:
> > > @ishitasequeira :
> > > 
> > > * I have rebased the PR against master
> > > * The lint failure seems to be caused by a caching issue. I don't have enough permissions to clear it though, can you take care of this ?
> > > 
> > > Antoine
> > 
> > 
> > I don't think I have write access to your branch. Can you provide me access to it?
> 
> I've sent you an invite

nvm, I got distracted by the tar errors in the `Lint Go code` step: I've fixed code formatting in 4e7d8669980b8b5f7e822b82da6b7eecb4fccaaf.
Code linting check is ok now.

I'm now facing an error in end-to-end tests for `TestNamespacedDisableManifestGeneration` but I'm not sure to understand what is the underlying problem.

Antoine

ishitasequeira commented:
/cherry-pick release-2.8
